### PR TITLE
Allow creation of an article

### PR DESCRIPTION
**Pull request by nlisgo**
Expose an endpoint to allow the creation of an article with the only field being name.

Create a repository for the articles in memory and list the articles on the article list end point.

### CODE REVIEW

File: test/errors/article-has-no-id.test.ts [hunk 1]
```diff
@@ -1,14 +1,14 @@
-import ArticleHasNoId from '../../src/errors/article-has-no-id';
+import ArticleHasNoID from '../../src/errors/article-has-no-id';
```
Comment (thewilkybarkid) [inline test/errors/article-has-no-id.test.ts]: Don't disagree, but reducing the diff in this PR.

File: src/routes/post-article.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+import { JsonLdObj } from 'jsonld/jsonld-spec';
+import uniqueString from 'unique-string';
+import Articles from '../articles';
+import ArticleIdAlreadySet from '../errors/article-id-already-set';
+
+export default (articles: Articles): Middleware => (
+  async ({ request, response }: Context, next: Next): Promise<void> => {
+    const article = request.body as JsonLdObj;
+    if ('@id' in article) {
+      throw new ArticleIdAlreadySet(article['@id']);
```
Comment (thewilkybarkid) [inline src/routes/post-article.ts]: Been thinking about this. `409 Conflict` makes sense if an article with that ID is already in the collection, but really we're saying you're not allowed to set an ID, so `403 Forbidden` covers it. Suggest we also throw an HTTP error directly here for simplicity.

Comment (nlisgo) [reply src/routes/post-article.ts]: I've made this refactor.

File: src/routes/post-article.ts [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+import { JsonLdObj } from 'jsonld/jsonld-spec';
+import uniqueString from 'unique-string';
+import createHttpError from 'http-errors';
+import Articles from '../articles';
+
+export default (articles: Articles): Middleware => (
+  async ({ request, response }: Context, next: Next): Promise<void> => {
+    const article = request.body as JsonLdObj;
+    if ('@id' in article) {
+      throw new createHttpError.Forbidden('Article ID is not permitted');
```
Comment (nlisgo) [inline src/routes/post-article.ts]: Is there value is passing the id into the message for logging?

Perhaps: `Article ID is not permitted ('_:1' given)`

File: src/routes/index.ts [hunk 4]
```diff
@@ -1,6 +1,7 @@
 enum Routes {
   'ApiDocumentation' = 'api-documentation',
   'ArticleList' = 'article-list',
+  'PostArticle' = 'post-article',
```
Comment (thewilkybarkid) [inline src/routes/index.ts]: Suggest we call it `CreateArticle`. Maybe even `AddArticle`?

Comment (nlisgo) [reply src/routes/index.ts]: Will go for `AddArticle` as it matches the method call.

File: src/routes/post-article.ts (hunk 3 above)
Comment (thewilkybarkid) [reply src/routes/post-article.ts]: Yeah, maybe even `Article IDs must not be set ('_:1' was given)`.

**Comment by thewilkybarkid** [comment]
I've added 3 tasks to https://github.com/libero/publisher/issues/291 (re validation), ok to not include in this PR (matches the API documentation) but we'll need a follow-up.

File: src/routes/post-article.ts [hunk 5]
```diff
@@ -0,0 +1,22 @@
+import { constants } from 'http2';
+import { Context, Middleware, Next } from 'koa';
+import { JsonLdObj } from 'jsonld/jsonld-spec';
+import uniqueString from 'unique-string';
+import createHttpError from 'http-errors';
+import Articles from '../articles';
+
+export default (articles: Articles): Middleware => (
+  async ({ request, response }: Context, next: Next): Promise<void> => {
+    const article = request.body as JsonLdObj;
```
Comment (thewilkybarkid) [inline src/routes/post-article.ts]: Really we'll want the context to know that the body is JSON-LD. We can follow this up.

**Review by thewilkybarkid — APPROVED** [review]

**Pull request merged**